numba/cuda/simulator/cudadrv/devicearray.py

```diff
@@ -30,23 +30,54 @@ class FakeShape(tuple):
     indexing, similar to the shape in CUDA Python. (Numpy shape arrays allow
     negative indexing)
     '''
+
     def __getitem__(self, k):
         if isinstance(k, int) and k < 0:
             raise IndexError('tuple index out of range')
         return super(FakeShape, self).__getitem__(k)
 
 
+class FakeKernelCUDAArray(object):
+    def __init__(self, item):
+        assert isinstance(item, FakeCUDAArray)
+        self.__dict__['_item'] = item
+
+    def __wrap_if_fake(self, item):
+        return FakeKernelCUDAArray(item) if isinstance(item, FakeCUDAArray) else item
+
+    def __getattr__(self, attrname):
+        if attrname in dir(self.__dict__['_item']._ary):  # For, eg, array size.
+            return self.__wrap_if_fake(getattr(self.__dict__['_item']._ary, attrname))
+        else:
+            return self.__wrap_if_fake(self._item.__getitem__(attrname))
+
+    def __setattr__(self, nm, val):
+        self._item.__setitem__(nm, val)
+
+    def __getitem__(self, idx):
+        return self.__wrap_if_fake(self._item.__getitem__(idx))
+
+    def __setitem__(self, idx, val):
+        self._item.__setitem__(idx, val)
+
+    def __len__(self):
+        return len(self.__dict__['_item'])
+
 class FakeCUDAArray(object):
     '''
     Implements the interface of a DeviceArray/DeviceRecord, but mostly just
     wraps a NumPy array.
     '''
 
-    __cuda_ndarray__ = True     # There must be gpu_data attribute
-
+    __cuda_ndarray__ = True  # There must be gpu_data attribute
 
     def __init__(self, ary, stream=0):
-        self._ary = ary.reshape(1) if ary.ndim == 0 else ary
+        if ary.ndim == 0:
```

Comment:
Can you add a comment explaining why this case is needed?

Reply:
Done.